Additional checks for input existence in transactions

piotr-iohk opened the pull request:
# Issue Number

It turns out there are no checks for that.
A bit related to #2327.

# Overview

- 2605a9f1071aad41bca21d8ef98210c117f7cfb9
  Additional checks for inputs in incoming and outgoing transactions in integration tests
  
- 1247f9a75598a3b3f44d6b7631645868fac272a4
  Additional checks for inputs in join/quit pool transactions
  
- bd2916b0423811b85328edb60ae427f2e4b4dee4
  remove redundant cases & fix fixture wallet sharing common funding UTxOs
      It's a bit fucked up. Since the scenario is using fixtureWallet, the
    UtxO from the wallet comes from one of the fixture transaction made
    when initializing the cluster. There are many pre-funded fixture
    wallets with 10 UTxO, but in order not to spend 10 days funding them
    all, the cluster generate pretty large transaction with ~100
    outputs. So a single fixture transaction funds about 10 fixture
    wallets.

    Now it means that, a particular fixture wallet will know the value
    of all inputs of its funding fixture transaction and therefore, when
    picking two fixtureWallet one after the other, it may likely happen
    that both wallets are able to resolve inputs of any UTxO coming from
    their original fixture transaction.

    I fixed it by simply using fixtureWalletWith [...] which generates
    completly new and uncorrelated UTxOs.

- cdf556ac20da1ca7afe12c831a0876633f704317
  Add input check for multi-address pending tx



# Comments

<!-- Additional comments or screenshots to attach if any -->

<!--
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Jira will detect and link to this PR once created, but you can also link this PR in the description of the corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
 ✓ Finally, in the PR description delete any empty sections and all text commented in <!--, so that this text does not appear in merge commit messages.
-->

Commit:
Additional checks for inputs in incoming and outgoing transactions in integration tests

Commit:
Additional checks for inputs in join/quit pool transactions

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -215,12 +217,32 @@ spec = describe "SHELLEY_STAKE_POOLS" $ do
+            -- all tx inputs have address and amount
+            , expectField #inputs $ \inputs' -> do
+                let addrAmt = (fmap source inputs')
+                addrAmt `shouldSatisfy` (all isJust)
```
Test fails here... it turns out that for delegation inputs are not there in the output of the join/quit transaction.

piotr-iohk replied:
@KtorZ as far as your simplifications to the checks, I see you have removed this one and also the one from multi-output transactions. I'd argue that these are redundant, especially that one above was failing before (not because of `fixtureWallet` thing). It is OK now (after https://github.com/input-output-hk/cardano-wallet/pull/2370 I suppose) but still worth checking.

piotr-iohk replied:
Whoops, sorry, you didn't remove that. It was the commit diff that mislead me :sweat_smile:
Anyway... I have added small additional check to multi-output tx in cdf556a.

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:
```diff
@@ -289,17 +290,55 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
+                -- as the tx is incoming
+                , expectField #inputs $ \inputs' -> do
+                    let addrAmt2 = (fmap source inputs')
+                    addrAmt2 `shouldSatisfy` (all isNothing)
```
Fails here :thinking: ... inputs are there on incoming transaction on `fixtureWallet`

piotr-iohk replied:
I can confirm this also on `shelley-test-cluster`. So it seems that:
 -  in case of transaction `fixtureWalletSrc` -> `fixtureWalletDest` there are resolved inputs visible on `incoming` transaction on `fixtureWalletDest`.
 - in case of transaction `emptyWallet` -> `fixtureWallet` there are NO resolved inputs visible on `incoming` transaction on `fixtureWallet`

Could this be due to fixtureWallet share the same key or something, so they are able to resolve the inputs in case of transaction between them?

KtorZ replied:
Yeah it's a bit fucked up. Since the scenario is using `fixtureWallet`, the UtxO from the wallet comes from one of the fixture transaction made when initializing the cluster. There are many pre-funded fixture wallets with 10 UTxO, but in order not to spend 10 days funding them all, the cluster generate pretty large transaction with ~100 outputs. So a single fixture transaction funds about 10 fixture wallets. 

Now it means that, a particular fixture wallet will know the value of all inputs of its funding fixture transaction and therefore, when picking two fixtureWallet one after the other, it _may_ likely happen that both wallets are able to resolve inputs of any UTxO coming from their original fixture transaction. 

I fixed it by simply using `fixtureWalletWith [...]` which generates completly new and uncorrelated UTxOs.

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:
```diff
@@ -357,15 +396,55 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
+                -- as the tx is incoming
+                , expectField #inputs $ \inputs' -> do
+                    let addrAmt2 = (fmap source inputs')
+                    addrAmt2 `shouldSatisfy` (all isNothing)
```
Doesn't fail here... The difference is that the destination wallet is `emptyWallet`.

KtorZ committed:
remove redundant cases & fix fixture wallet sharing common funding UTxOs

    It's a bit fucked up. Since the scenario is using fixtureWallet, the
    UtxO from the wallet comes from one of the fixture transaction made
    when initializing the cluster. There are many pre-funded fixture
    wallets with 10 UTxO, but in order not to spend 10 days funding them
    all, the cluster generate pretty large transaction with ~100
    outputs. So a single fixture transaction funds about 10 fixture
    wallets.

    Now it means that, a particular fixture wallet will know the value
    of all inputs of its funding fixture transaction and therefore, when
    picking two fixtureWallet one after the other, it may likely happen
    that both wallets are able to resolve inputs of any UTxO coming from
    their original fixture transaction.

    I fixed it by simply using fixtureWalletWith [...] which generates
    completly new and uncorrelated UTxOs.

KtorZ reviewed (approved):
Reviewed it, fixed it and removed some redundant cases :+1:

Commit:
Add input check for multi-address pending tx
